homeassistant/components/rfxtrx/__init__.py

```diff
@@ -235,16 +243,29 @@ def handle_receive(event):
 
         _LOGGER.debug("Receive RFXCOM event: %s", event_data)
 
-        data_bits = get_device_data_bits(event.device, device_bits)
+        data_bits = get_device_data_bits(event.device, devices)
         device_id = get_device_id(event.device, data_bits=data_bits)
 
         # Callback to HA registered components.
         hass.helpers.dispatcher.dispatcher_send(SIGNAL_EVENT, event, device_id)
 
         # Signal event to any other listeners
-        if device_id in device_events:
+        fire_event = devices.get(device_id, {}).get(CONF_FIRE_EVENT)
+        if fire_event:
             hass.bus.fire(EVENT_RFXTRX_EVENT, event_data)
 
+    @callback
+    def device_update(event, device_id):
+        if device_id not in devices:
+            data = entry.data.copy()
+            event_code = binascii.hexlify(event.data).decode("ASCII")
+            data[CONF_DEVICES][event_code] = device_id
+            hass.config_entries.async_update_entry(entry=entry, data=data)
+            devices[device_id] = {}
+
+    if config[CONF_AUTOMATIC_ADD]:
+        hass.helpers.dispatcher.async_dispatcher_connect(SIGNAL_EVENT, device_update)
```

Comment:
I think we should move the call off this function into handle_event() instead. Not connect it by dispatcher. That way we know it's finished before any platform get's data for the event.

Reply:
I have a branch in progress: https://github.com/home-assistant/core/compare/dev...elupus:rftrx_async?expand=1 where handle_event ends up async. that way can stay in async context pretty much everywhere and avoid thread or race things.

Reply:
The platforms do not depend on the data that is being updated inside device_update, so I think it doesn't really matter

Reply:
As long as multiple device_updates do not run in parallel, we should be fine here

Reply:
One semi issue is that the variable devices[] will now be accessed both by event loop and thread executors. Given pythons GIL lock. It won't cause an issue. But it's ugly :)

Reply:
Perhaps I remove the refactor then, address it later, then the devices is only used in the event loop. I don't wanna block this PR on such a thing, because I mainly need this to get progress on the options flow

Reply:
I'm ok with this.

Reply:
We missed one thing here. Can't use async_dispatcher_connect here since setup_internal is not event loop context. I will fix in my followup.

Reply:
Yes! Probably best to change the `setup_internal` context to async and schedule to the executor from that when needed. Not very much seem to do I/O besides the callbacks that are just registered.